[JENKINS-69339] Add support for badge icons in Management links

mawinter69 opened the pull request:
Allow to define a badge icon and a tooltip for all management links.

Adjust PluginsLink to the new method
Add implementations for OldDataMonitor and NodesLink

See [JENKINS-69339](https://issues.jenkins.io/browse/JENKINS-69339).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Add support for badge icons in Management links

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.
The changelog entry should be in the imperative mood; e.g., write "do this"/"return that" rather than "does this"/"returns that".
For examples, see: https://www.jenkins.io/changelog/
-->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content-Security-Policy directives (see [documentation on jenkins.io](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6995"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

mawinter69 committed:
[JENKINS-69339] badge icon for Management links

Allow to define a badge icon and a tooltip for all management links

daniel-beck commented on core/src/main/java/hudson/model/ManagementLink.java:
```diff
@@ -214,5 +214,26 @@ public enum Category {
+     * @return badge text or {@code null} if no badge should be shown.
+     * @since TODO
+     */
+    public @CheckForNull String getBadge() {
+        return null;
+    }
```
This should probably be limited to returning numbers, or is there a real use case for non-numeric badge values that wouldn't quickly overwhelm the badge size?

mawinter69 replied:
I was thinking for my agent maintenance plugin to display something like 5/12 with a tooltip "5 active maintenance windows 12 total"

mawinter69 replied:
Another use case might be to show a "!" to indicate that some action needs to be taken

daniel-beck replied:
Interesting, good point with 5/12.

Still, I wonder whether we could define this API so that it prevents weird abuse (like actual text, as happened in `groovy-postbuild` with build badges)

mawinter69 replied:
when returning 1000000 it will probably also look weird. One could also think of an array if integers and take only the first 2 or 3 elements of the array and separate it with "/".
Or just take a substring of the first 8 to 10 characters.

mawinter69 replied:
Some samples how it would look:

Long string without blanks
![image](https://user-images.githubusercontent.com/17767050/184668112-85340587-29d1-4ec4-97b4-13b4cff29e46.png)

String with blanks leads to a line break
![image](https://user-images.githubusercontent.com/17767050/184668200-4f703b07-3ccb-4175-8abf-952e33bc49f5.png)

![image](https://user-images.githubusercontent.com/17767050/184668334-082ec05a-e18e-4210-bc8c-59cd6228aa6a.png)

![image](https://user-images.githubusercontent.com/17767050/184668582-da42070e-65c2-4550-82e6-60e30efdacdb.png)

daniel-beck replied:
> when returning 1000000 it will probably also look weird.

Realistically, no number will grow that large. The upper bounds should be hundreds of plugins need updating, (tens of) thousands of jobs fail to load, so that's ~5 digits?

daniel-beck replied:
Also, if it's a string you'll need to define whether it supports HTML or not. Right now it would (and will need to escape any user content).

daniel-beck commented on core/src/main/java/hudson/model/ManagementLink.java:
```diff
@@ -214,5 +214,26 @@ public enum Category {
+
+    /**
+     * The tooltip to show for the badge on "Manage Jenkins".
+     * When a badge is shown this should not be an empty string.
+     *
+     * @return badge tooltip
+     * @since TODO
+     */
+    public @NonNull String getBadgeTooltip() {
+        return "";
```
Seems weird to require an empty string when there's no content. Why?

mawinter69 committed:
allow null for the tooltip

fix compile errors

daniel-beck commented:
The weird thing about the existing badge is that it only exists for a feature that does not have a corresponding admin monitor (plugin updates). My guess is that it's probably deliberately so, because there are always plugin updates, especially since JEP-229, so having an admin monitor that's always on would be spammy and quickly ignored by many admins.

Other features like Old Data Monitor do have an admin monitor when they have useful content. Usually because this is something that _needs_ attention (in that case, loss of configuration data, loss of functionality, etc.).

If this is added, we need to provide guidelines to explain when you would have the admin monitor, when you would have the badge, and when you would have both.

An interesting case is In-Process Script Approval (`script-security` plugin): It currently has a badge-like feature (the label changes). This feature should be migrated to badges if this is merged (would make a nice downstream PR for this change), and I wonder whether an admin monitor would make sense there as well. Could probably go either way.

basil committed:
Merge branch 'master' into JENKINS-69339

basil committed:
Merge branch 'master' into JENKINS-69339

daniel-beck commented on core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:
```diff
@@ -85,9 +85,9 @@ THE SOFTWARE.
-                        <div tooltip="${m.getUpdateCount() == 1 ? '%updateAvailable(m.getUpdateCount())' : '%updatesAvailable(m.getUpdateCount())'}" class="jenkins-section__item__icon__badge">
-                          ${m.getUpdateCount()}
+                      <j:if test="${m.badge != null}">
+                        <div tooltip="${m.badgeTooltip}" class="jenkins-section__item__icon__badge">
```
This is now (once re-merged with `master`) a non-HTML tooltip. That seems fine.

daniel-beck committed:
Merge branch 'master' into JENKINS-69339

daniel-beck reviewed (commented):
I really like this PR. It cleans up the existing tooltip for plugin updates and adds useful additional tooltips (although I'm not entirely sure the node one is useful as-is, given agents on a schedule etc.).

An open question remains about the guidelines for plugin developers on how to use this feature. We should have documentation when to (not) use this ready at the same time we publish this API addition. If we keep the `String` type, the docs should also explain it's for short labels only (probably mostly numeric; e.g., `4 of 12` might be OK, `21 available updates` is not). This should be added to Javadoc and/or https://plugins.jenkins.io/design-library/

Otherwise (and assuming a quick re-test confirms the non-HTML tooltip behavior) this looks good 👍

daniel-beck commented:
> assuming a quick re-test confirms the non-HTML tooltip behavior

Confirmed, non-HTML tooltip and content when merged with `master`.

mawinter69 commented:
I'm wondering if there are other places where such badges might be useful. If so it might make sense to create a Badge class that contains the data, tooltip and maybe a severity so the badge can have different colors. Even when not used somewhere else it might help to avoid running some calculations twice (once for the badge and once for the tooltip.
For severity, e.g. when there are plugin updates that fix security issues display it in red, otherwise in orange.

mawinter69 commented:
> (although I'm not entirely sure the node one is useful as-is, given agents on a schedule etc.).

The agents on a schedule wouldn't be a problem, but the ondemand would show up whenever they have nothing to do and thus disconnect. Maybe restrict the check to agents that use the always online strategy which is the default and most people use I guess.

daniel-beck commented:
> avoid running some calculations twice (once for the badge and once for the tooltip

For expensive calculations, could be handled by individual `ManagementLink` in a cache. I don't think there are _that_ many cases where computation would be expensive?

Feel free to sketch this out, might be reasonable for future extension (e.g. severity/color), but IMO all that's missing here is related documentation.

> Maybe restrict the check to agents that use the always online strategy which is the default and most people use I guess.

Right, something along those lines. Might be simpler to remove this from the PR for now and look into it as followup. Migrating Manage Plugins to this is already pretty nice cleanup, and Manage Old Data is a reasonable new use.

mawinter69 committed:
create Badge class

Badge class takes message, tooltip and severity

Do not create badge for NodesLink. It will not work with the
agent-maintenance plugin as this configures as well a retention strategy
that can't be queried here. To get the reliably working Demand and
SimpleScheduled would need some reworking at least.

The plugin link badge now changes color when there is an update that
fixes a security vulnerability.

mawinter69 committed:
minor fixes to javadoc and whitespace

mawinter69 committed:
set badge var just before we use it

mawinter69 commented:
I removed the badge for the nodes now, there is no way of reliably determining this information, e.g. when the agent-maintenance enabled for an agent, which is proxying the retentionstrategy 
Added a Badge class to wrap all the information
Plugin update badge is now red when updates fix security vulnerabilities and orange otherwise. Tooltip shows the actual numbers for total updates and updates with security fixes. One could think of extending the badge/tooltip also for plugins that have incompatible changes

mawinter69 committed:
remove text color for badge item

daniel-beck commented on core/src/main/java/hudson/model/ManagementLink.java:
```diff
@@ -214,5 +215,15 @@ public enum Category {
+     * @return badge or {@code null} if no badge should be shown.
+     * @since TODO
+     */
+    public @CheckForNull Badge getBadge() {
```
One drawback of the new `Badge` type is that it requires a new core dependency. Previously, the `String` based API could be implemented even by plugins with an old core dependency.

I don't think this change must be reverted, but considering simple page navigation shouldn't be expensive anyway, the argument for one call instead of two isn't _that_ strong (expensive computation should be done elsewhere and cached values shown, kind of like available updates).

daniel-beck commented on core/src/main/java/jenkins/management/PluginsLink.java:
```diff
@@ -72,13 +72,43 @@ public Category getCategory() {
+                    case 0:
+                        tooltip = Messages.PluginsLink_updatesAvailable(size);
+                        break;
+                    case 1:
+                        tooltip = Messages.PluginsLink_oneSecurityUpdateAvailable(size);
+                        severity = Badge.Severity.danger;
+                        break;
```
Interesting approach, I kinda like it. Less red is better when there's no immediate need to act, especially with a badge that is practically always visible.

FWIW I look forward to reading in docs about the interaction between admin monitors and badges here :)

daniel-beck commented on core/src/main/java/jenkins/management/Badge.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    public enum Severity {
+        danger, warning, info
```
Convention is for these to be ALLCAPS.

(I expected there to be a static check for this, kinda surprised.)

daniel-beck commented on core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:
```diff
@@ -85,9 +85,10 @@ THE SOFTWARE.
-                          ${m.getUpdateCount()}
+                      <j:set var="badge" value="${m.badge}"/>
+                      <j:if test="${badge != null}">
+                        <div tooltip="${badge.tooltip}" class="jenkins-section__item__icon__badge alert-${badge.severity}">
```
Does this handle `null` tooltips correctly? (i.e., no tooltip) There's no `@NonNull` annotation so I expect them to be optional.

(Untested, planning to check later unless someone else is faster.)

timja replied:
The constructor for badge has nonnull on tooltip now

daniel-beck commented on core/src/main/resources/hudson/model/ManageJenkinsAction/index.jelly:
```diff
@@ -85,9 +85,10 @@ THE SOFTWARE.
-                          ${m.getUpdateCount()}
+                      <j:set var="badge" value="${m.badge}"/>
+                      <j:if test="${badge != null}">
+                        <div tooltip="${badge.tooltip}" class="jenkins-section__item__icon__badge alert-${badge.severity}">
```
No `@NonNull` annotation on `severity`, and no fallback handling for `null` parameter, but I don't think the result looks reasonable when there's none.

(Untested, planning to check later unless someone else is faster.)

timja replied:
severity is now non null

daniel-beck commented on war/src/main/less/modules/section.less:
```diff
@@ -103,14 +103,12 @@
       display: flex;
       align-items: center;
       justify-content: center;
-      background: #ff0b27;
       border-radius: 100px;
       font-size: 0.7rem;
       min-height: 20px;
       min-width: 20px;
       padding: 0 0.4rem;
       box-shadow: 0 1px 1px rgba(black, 0.1);
-      color: white;
```
Does this only affect Manage Jenkins badges? `.jenkins-section` looks very generic.

mawinter69 committed:
improve documentation

add info about incompatible changes

timja committed:
Merge branch 'master' into JENKINS-69339

timja commented on core/src/main/java/jenkins/management/Badge.java:
```diff
@@ -0,0 +1,103 @@
+     * Create a new Badge
+     *
+     * @param text  The text to be shown in the badge.
+     *      Keep it short, ideally just a number. More than 6 or 7 characters do not look good. Do not put in blanks
```
what do you mean by blanks?

mawinter69 replied:
A whitespace can lead to a line break getting rendered. See the first outdated conversation where I added some screenshots

timja replied:
what about?

```suggestion
     *      Keep it short, ideally just a number. More than 6 or 7 characters do not look good. Avoid spaces as they will lead to line breaks.
```

timja committed:
Copyediting

timja reviewed (approved):
LGTM just one wording question in the Javadoc but no blockers, tested locally.

timja committed:
Merge branch 'master' into JENKINS-69339

mawinter69 committed:
Update core/src/main/java/jenkins/management/Badge.java

Co-authored-by: Tim Jacomb <21194782+timja@users.noreply.github.com>

timja commented:
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

NotMyFault committed:
Merge branch 'master' into JENKINS-69339

daniel-beck commented:
> [timja](https://github.com/timja) dismissed [daniel-beck](https://github.com/daniel-beck)’s [stale review](https://github.com/jenkinsci/jenkins/pull/6995#pullrequestreview-1223585941) Dec 31, 2022
> addressed

In multiple comments I kept requesting developer documentation. Could you point me to it?

https://github.com/jenkinsci/jenkins/pull/6995#issuecomment-1216306653:

> If this is added, we need to provide guidelines to explain when you would have the admin monitor, when you would have the badge, and when you would have both.

https://github.com/jenkinsci/jenkins/pull/6995#pullrequestreview-1219211363:

> An open question remains about the guidelines for plugin developers on how to use this feature. We should have documentation when to (not) use this ready at the same time we publish this API addition. If we keep the `String` type, the docs should also explain it's for short labels only (probably mostly numeric; e.g., `4 of 12` might be OK, `21 available updates` is not). This should be added to Javadoc and/or https://plugins.jenkins.io/design-library/

https://github.com/jenkinsci/jenkins/pull/6995#pullrequestreview-1223585941:

> FWIW I look forward to reading in docs about the interaction between admin monitors and badges here :)

timja commented:
> In multiple comments I kept requesting developer documentation. Could you point me to it?

https://github.com/jenkinsci/jenkins/blob/a6d542509799d1b70610406c7750f2440c44e8b2/core/src/main/java/jenkins/management/Badge.java#L33-L50

https://github.com/jenkinsci/jenkins/blob/a6d542509799d1b70610406c7750f2440c44e8b2/core/src/main/java/hudson/model/ManagementLink.java#L218-L223

daniel-beck commented:
Thanks, missed it (or forgot in the mean time) 👍
